File: lru-cache.java
class DListNode {
    public int key;
    public int value;
    public DListNode prev;
    public DListNode next;
    
    public DListNode(int key, int value) {
        this.key = key;
        this.value = value;
        this.prev = null;
        this.next = null;
    }
}

class LRUCache {
    
    private int capacity;
    private DListNode head = new DListNode(-1, -1);
    private DListNode tail = new DListNode(-1, -1);
    private HashMap<Integer, DListNode> hashmap = new HashMap<>();

    public LRUCache(int capacity) {
        this.capacity = capacity;
        head.next = tail;
        tail.prev = head;
    }

    public int get(int key) {
        if (!hashmap.containsKey(key)) {
            return -1;
        }
        DListNode node = hashmap.get(key);
        node.prev.next = node.next;
        node.next.prev = node.prev;

        node.next = tail;
        node.prev = tail.prev;
        tail.prev.next = node;
        tail.prev = node;
        return node.value;
    }

    public void put(int key, int value) {
        if (get(key) != -1) {
            DListNode node = hashmap.get(key);
            node.value = value;
            return;
        }
        
        if (hashmap.size() == capacity) {
            hashmap.remove(head.next.key);
            head.next = head.next.next;
            head.next.prev = head;
        }
        
        DListNode node = new DListNode(key, value);
        hashmap.put(key, node);
        node.next = tail;
        node.prev = tail.prev;
        tail.prev.next = node;
        tail.prev = node;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: lru-cache.py
class DLNode(object):
    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None

class LRUCache:

    def __init__(self, capacity: int):
        self.table = {}
        self.capacity = capacity
        self.head = DLNode(-1, -1)
        self.tail = DLNode(-1, -1)
        self.head.next = self.tail
        self.tail.prev = self.head

    def get(self, key: int) -> int:
        if key not in self.table:
            return -1
        
        node = self.table[key]
        
        # remove node from DL list
        node.prev.next = node.next
        node.next.prev = node.prev
        
        # attach to the front
        node.prev = self.head
        node.next = self.head.next
        node.next.prev = node
        self.head.next = node
        return node.val

    def put(self, key: int, value: int) -> None:
        if self.get(key) != -1:
            node = self.table[key]
            node.val = value
            return
    
        if len(self.table) == self.capacity:
            leastUsedNode = self.tail.prev
            del self.table[leastUsedNode.key]
            leastUsedNode.prev.next = self.tail
            self.tail.prev = leastUsedNode.prev
        
        node = DLNode(key, value)
        self.table[key] = node
        node.prev = self.head
        node.next = self.head.next
        self.head.next.prev = node
        self.head.next = node
        return


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
